File: raytracer/raytracer/tests/matrix-tests.cpp
#ifdef TEST_BUILD

#include "Catch.h"
#include "math/matrix.h"
#include "math/transformation-matrices.h"

using namespace math;
using namespace math::transformation_matrices;


TEST_CASE("[Matrix] Matrix transformation identity<4>() equals identity<4>()", "[Matrix]")
{
    auto a = identity<4>();
    auto b = identity<4>();

    REQUIRE( a == approx(b) );
}

TEST_CASE("[Matrix] Matrix transformation translation(Vector3D(0,0,0)) equals identity<4>()", "[Matrix]")
{
    auto a = translation(Vector3D(0,0,0));
    auto b = identity<4>();

    REQUIRE( a == approx(b) );
}

TEST_CASE("[Matrix] Matrix transformation scaling(1,1,1) equals identity<4>()", "[Matrix]")
{
    auto a = scaling(1,1,1);
    auto b = identity<4>();

    REQUIRE( a == approx(b) );
}

TEST_CASE("[Matrix] Matrix transformation rotation_around_x(0_degrees) equals identity<4>()", "[Matrix]")
{
    auto a = rotation_around_x(0_degrees);
    auto b = identity<4>();

    REQUIRE( a == approx(b) );
}

TEST_CASE("[Matrix] Matrix transformation rotation_around_y(0_degrees) equals identity<4>()", "[Matrix]")
{
    auto a = rotation_around_y(0_degrees);
    auto b = identity<4>();

    REQUIRE( a == approx(b) );
}

TEST_CASE("[Matrix] Matrix transformation rotation_around_z(0_degrees) equals identity<4>()", "[Matrix]")
{
    auto a = rotation_around_z(0_degrees);
    auto b = identity<4>();

    REQUIRE( a == approx(b) );
}

TEST_CASE("[Matrix] Matrix transformation rotation_around_x(0_degrees) equals rotation_around_x(360_degrees)", "[Matrix]")
{
    auto a = rotation_around_x(0_degrees);
    auto b = rotation_around_x(360_degrees);

    REQUIRE( a == approx(b) );
}

TEST_CASE("[Matrix] Matrix transformation rotation_around_x(-90_degrees) equals rotation_around_x(270_degrees)", "[Matrix]")
{
    auto a = rotation_around_x(-90_degrees);
    auto b = rotation_around_x(270_degrees);

    REQUIRE( a == approx(b) );
}

TEST_CASE("[Matrix] Matrix transformation rotation_around_x(-180_degrees) equals rotation_around_x(180_degrees)", "[Matrix]")
{
    auto a = rotation_around_x(-180_degrees);
    auto b = rotation_around_x(180_degrees);

    REQUIRE( a == approx(b) );
}

TEST_CASE("[Matrix] Matrix transformation scaling(-1,1,-1) equals rotation_around_y(180_degrees)", "[Matrix]")
{
    auto a = scaling(-1,1,-1);
    auto b = rotation_around_y(180_degrees);

    REQUIRE( a == approx(b) );
}

TEST_CASE("[Matrix] Matrix transformation scaling(-1,-1,1) equals rotation_around_z(180_degrees)", "[Matrix]")
{
    auto a = scaling(-1,-1,1);
    auto b = rotation_around_z(180_degrees);

    REQUIRE( a == approx(b) );
}

TEST_CASE("[Matrix] Matrix transformation scaling(1,-1,-1) equals rotation_around_x(180_degrees)", "[Matrix]")
{
    auto a = scaling(1,-1,-1);
    auto b = rotation_around_x(180_degrees);

    REQUIRE( a == approx(b) );
}

TEST_CASE("[Matrix] Matrix multiplication of identity<4>() and identity<4>()", "[Matrix]")
{
    auto a = identity<4>();
    auto b = identity<4>();
    auto actual = a * b;
    auto expected = identity<4>();

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of identity<4>() and scaling(2, 1, 1)", "[Matrix]")
{
    auto a = identity<4>();
    auto b = scaling(2, 1, 1);
    auto actual = a * b;
    auto expected = scaling(2, 1, 1);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of identity<4>() and scaling(1, 3, 1)", "[Matrix]")
{
    auto a = identity<4>();
    auto b = scaling(1, 3, 1);
    auto actual = a * b;
    auto expected = scaling(1, 3, 1);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of identity<4>() and scaling(1, 1, 4)", "[Matrix]")
{
    auto a = identity<4>();
    auto b = scaling(1, 1, 4);
    auto actual = a * b;
    auto expected = scaling(1, 1, 4);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of identity<4>() and rotation_around_x(40_degrees)", "[Matrix]")
{
    auto a = identity<4>();
    auto b = rotation_around_x(40_degrees);
    auto actual = a * b;
    auto expected = rotation_around_x(40_degrees);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of identity<4>() and rotation_around_y(10_degrees)", "[Matrix]")
{
    auto a = identity<4>();
    auto b = rotation_around_y(10_degrees);
    auto actual = a * b;
    auto expected = rotation_around_y(10_degrees);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of identity<4>() and rotation_around_z(250_degrees)", "[Matrix]")
{
    auto a = identity<4>();
    auto b = rotation_around_z(250_degrees);
    auto actual = a * b;
    auto expected = rotation_around_z(250_degrees);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of scaling(2, 1, 1) and scaling(2, 1, 1)", "[Matrix]")
{
    auto a = scaling(2, 1, 1);
    auto b = scaling(2, 1, 1);
    auto actual = a * b;
    auto expected = scaling(4, 1, 1);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of scaling(2, 1, 1) and scaling(0.5, 1, 1)", "[Matrix]")
{
    auto a = scaling(2, 1, 1);
    auto b = scaling(0.5, 1, 1);
    auto actual = a * b;
    auto expected = scaling(1, 1, 1);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of scaling(2, 1, 1) and scaling(1, 2, 1)", "[Matrix]")
{
    auto a = scaling(2, 1, 1);
    auto b = scaling(1, 2, 1);
    auto actual = a * b;
    auto expected = scaling(2, 2, 1);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of scaling(2, 1, 1) and scaling(1, 1, 2)", "[Matrix]")
{
    auto a = scaling(2, 1, 1);
    auto b = scaling(1, 1, 2);
    auto actual = a * b;
    auto expected = scaling(2, 1, 2);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of scaling(1, 3, 1) and scaling(1, 1, 2)", "[Matrix]")
{
    auto a = scaling(1, 3, 1);
    auto b = scaling(1, 1, 2);
    auto actual = a * b;
    auto expected = scaling(1, 3, 2);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_x(10_degrees) and rotation_around_x(10_degrees)", "[Matrix]")
{
    auto a = rotation_around_x(10_degrees);
    auto b = rotation_around_x(10_degrees);
    auto actual = a * b;
    auto expected = rotation_around_x(20_degrees);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_x(10_degrees) and rotation_around_x(-10_degrees)", "[Matrix]")
{
    auto a = rotation_around_x(10_degrees);
    auto b = rotation_around_x(-10_degrees);
    auto actual = a * b;
    auto expected = identity<4>();

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_x(10_degrees) and rotation_around_x(20_degrees)", "[Matrix]")
{
    auto a = rotation_around_x(10_degrees);
    auto b = rotation_around_x(20_degrees);
    auto actual = a * b;
    auto expected = rotation_around_x(30_degrees);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_y(10_degrees) and rotation_around_y(20_degrees)", "[Matrix]")
{
    auto a = rotation_around_y(10_degrees);
    auto b = rotation_around_y(20_degrees);
    auto actual = a * b;
    auto expected = rotation_around_y(30_degrees);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_z(10_degrees) and rotation_around_z(20_degrees)", "[Matrix]")
{
    auto a = rotation_around_z(10_degrees);
    auto b = rotation_around_z(20_degrees);
    auto actual = a * b;
    auto expected = rotation_around_z(30_degrees);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of translation(Vector3D(1,0,0)) and translation(Vector3D(-1,0,0))", "[Matrix]")
{
    auto a = translation(Vector3D(1,0,0));
    auto b = translation(Vector3D(-1,0,0));
    auto actual = a * b;
    auto expected = identity<4>();

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of translation(Vector3D(0,5,0)) and translation(Vector3D(0,-5,0))", "[Matrix]")
{
    auto a = translation(Vector3D(0,5,0));
    auto b = translation(Vector3D(0,-5,0));
    auto actual = a * b;
    auto expected = identity<4>();

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of translation(Vector3D(0,0,9)) and translation(Vector3D(0,0,-9))", "[Matrix]")
{
    auto a = translation(Vector3D(0,0,9));
    auto b = translation(Vector3D(0,0,-9));
    auto actual = a * b;
    auto expected = identity<4>();

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of translation(Vector3D(1,0,0)) and translation(Vector3D(0,0,2))", "[Matrix]")
{
    auto a = translation(Vector3D(1,0,0));
    auto b = translation(Vector3D(0,0,2));
    auto actual = a * b;
    auto expected = translation(Vector3D(1,0,2));

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of translation(Vector3D(1,0,0)) and Vector3D(0,0,0)", "[Matrix]")
{
    auto a = translation(Vector3D(1,0,0));
    auto b = Vector3D(0,0,0);
    auto actual = a * b;
    auto expected = Vector3D(0,0,0);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of translation(Vector3D(0,1,0)) and Vector3D(0,0,0)", "[Matrix]")
{
    auto a = translation(Vector3D(0,1,0));
    auto b = Vector3D(0,0,0);
    auto actual = a * b;
    auto expected = Vector3D(0,0,0);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of translation(Vector3D(0,0,1)) and Vector3D(0,0,0)", "[Matrix]")
{
    auto a = translation(Vector3D(0,0,1));
    auto b = Vector3D(0,0,0);
    auto actual = a * b;
    auto expected = Vector3D(0,0,0);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of translation(Vector3D(1,0,0)) and Point3D(0,0,0)", "[Matrix]")
{
    auto a = translation(Vector3D(1,0,0));
    auto b = Point3D(0,0,0);
    auto actual = a * b;
    auto expected = Point3D(1,0,0);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of translation(Vector3D(0,1,0)) and Point3D(0,0,0)", "[Matrix]")
{
    auto a = translation(Vector3D(0,1,0));
    auto b = Point3D(0,0,0);
    auto actual = a * b;
    auto expected = Point3D(0,1,0);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of translation(Vector3D(0,0,1)) and Point3D(0,0,0)", "[Matrix]")
{
    auto a = translation(Vector3D(0,0,1));
    auto b = Point3D(0,0,0);
    auto actual = a * b;
    auto expected = Point3D(0,0,1);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of scaling(2,1,1) and Point3D(0,0,0)", "[Matrix]")
{
    auto a = scaling(2,1,1);
    auto b = Point3D(0,0,0);
    auto actual = a * b;
    auto expected = Point3D(0,0,0);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of scaling(2,1,1) and Point3D(1,0,0)", "[Matrix]")
{
    auto a = scaling(2,1,1);
    auto b = Point3D(1,0,0);
    auto actual = a * b;
    auto expected = Point3D(2,0,0);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of scaling(2,3,4) and Point3D(1,1,1)", "[Matrix]")
{
    auto a = scaling(2,3,4);
    auto b = Point3D(1,1,1);
    auto actual = a * b;
    auto expected = Point3D(2,3,4);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of scaling(2,3,4) and Point3D(2,2,2)", "[Matrix]")
{
    auto a = scaling(2,3,4);
    auto b = Point3D(2,2,2);
    auto actual = a * b;
    auto expected = Point3D(4,6,8);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_x(180_degrees) and Point3D(2,2,2)", "[Matrix]")
{
    auto a = rotation_around_x(180_degrees);
    auto b = Point3D(2,2,2);
    auto actual = a * b;
    auto expected = Point3D(2,-2,-2);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_y(180_degrees) and Point3D(2,2,2)", "[Matrix]")
{
    auto a = rotation_around_y(180_degrees);
    auto b = Point3D(2,2,2);
    auto actual = a * b;
    auto expected = Point3D(-2,2,-2);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_z(180_degrees) and Point3D(2,2,2)", "[Matrix]")
{
    auto a = rotation_around_z(180_degrees);
    auto b = Point3D(2,2,2);
    auto actual = a * b;
    auto expected = Point3D(-2,-2,2);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_x(180_degrees) and Vector3D(2,2,2)", "[Matrix]")
{
    auto a = rotation_around_x(180_degrees);
    auto b = Vector3D(2,2,2);
    auto actual = a * b;
    auto expected = Vector3D(2,-2,-2);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_y(180_degrees) and Vector3D(2,2,2)", "[Matrix]")
{
    auto a = rotation_around_y(180_degrees);
    auto b = Vector3D(2,2,2);
    auto actual = a * b;
    auto expected = Vector3D(-2,2,-2);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_z(180_degrees) and Vector3D(2,2,2)", "[Matrix]")
{
    auto a = rotation_around_z(180_degrees);
    auto b = Vector3D(2,2,2);
    auto actual = a * b;
    auto expected = Vector3D(-2,-2,2);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_x(90_degrees) and Point3D(0,0,0)", "[Matrix]")
{
    auto a = rotation_around_x(90_degrees);
    auto b = Point3D(0,0,0);
    auto actual = a * b;
    auto expected = Point3D(0,0,0);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_x(90_degrees) and Point3D(1,0,0)", "[Matrix]")
{
    auto a = rotation_around_x(90_degrees);
    auto b = Point3D(1,0,0);
    auto actual = a * b;
    auto expected = Point3D(1,0,0);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_x(90_degrees) and Point3D(0,1,0)", "[Matrix]")
{
    auto a = rotation_around_x(90_degrees);
    auto b = Point3D(0,1,0);
    auto actual = a * b;
    auto expected = Point3D(0,0,1);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_x(90_degrees) and Point3D(0,0,1)", "[Matrix]")
{
    auto a = rotation_around_x(90_degrees);
    auto b = Point3D(0,0,1);
    auto actual = a * b;
    auto expected = Point3D(0,-1,0);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_y(90_degrees) and Point3D(0,0,0)", "[Matrix]")
{
    auto a = rotation_around_y(90_degrees);
    auto b = Point3D(0,0,0);
    auto actual = a * b;
    auto expected = Point3D(0,0,0);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_y(90_degrees) and Point3D(0,1,0)", "[Matrix]")
{
    auto a = rotation_around_y(90_degrees);
    auto b = Point3D(0,1,0);
    auto actual = a * b;
    auto expected = Point3D(0,1,0);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_y(90_degrees) and Point3D(1,0,0)", "[Matrix]")
{
    auto a = rotation_around_y(90_degrees);
    auto b = Point3D(1,0,0);
    auto actual = a * b;
    auto expected = Point3D(0,0,-1);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_y(90_degrees) and Point3D(0,0,-1)", "[Matrix]")
{
    auto a = rotation_around_y(90_degrees);
    auto b = Point3D(0,0,-1);
    auto actual = a * b;
    auto expected = Point3D(-1,0,0);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_y(90_degrees) and Point3D(-1,0,0)", "[Matrix]")
{
    auto a = rotation_around_y(90_degrees);
    auto b = Point3D(-1,0,0);
    auto actual = a * b;
    auto expected = Point3D(0,0,1);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_z(90_degrees) and Point3D(0,0,0)", "[Matrix]")
{
    auto a = rotation_around_z(90_degrees);
    auto b = Point3D(0,0,0);
    auto actual = a * b;
    auto expected = Point3D(0,0,0);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_z(90_degrees) and Point3D(0,0,1)", "[Matrix]")
{
    auto a = rotation_around_z(90_degrees);
    auto b = Point3D(0,0,1);
    auto actual = a * b;
    auto expected = Point3D(0,0,1);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_z(90_degrees) and Point3D(1,0,0)", "[Matrix]")
{
    auto a = rotation_around_z(90_degrees);
    auto b = Point3D(1,0,0);
    auto actual = a * b;
    auto expected = Point3D(0,1,0);

    REQUIRE( actual == approx(expected) );
}

TEST_CASE("[Matrix] Matrix multiplication of rotation_around_z(90_degrees) and Point3D(0,1,0)", "[Matrix]")
{
    auto a = rotation_around_z(90_degrees);
    auto b = Point3D(0,1,0);
    auto actual = a * b;
    auto expected = Point3D(-1,0,0);

    REQUIRE( actual == approx(expected) );
}

#endif

File: raytracer/raytracer/test-generation/matrix-tests.rb
require './shared.rb'


test_file 'matrix-tests' do
  template do
    <<-'END'
      #ifdef TEST_BUILD

      #include "Catch.h"
      #include "math/matrix.h"
      #include "math/transformation-matrices.h"

      using namespace math;
      using namespace math::transformation_matrices;


      <<<TESTS>>>

      #endif
    END
  end

  test_suite do
    template do
      <<-END
        TEST_CASE("[Matrix] Matrix transformation #{first} equals #{second}", "[Matrix]")
        {
            auto a = #{first};
            auto b = #{second};

            REQUIRE( a == approx(b) );
        }
      END
    end

    test_case do |data|
      data.first = 'identity<4>()'
      data.second = 'identity<4>()'
    end

    test_case do |data|
      data.first = 'translation(Vector3D(0,0,0))'
      data.second = 'identity<4>()'
    end


    test_case do |data|
      data.first = 'scaling(1,1,1)'
      data.second = 'identity<4>()'
    end

    test_case do |data|
      data.first = 'rotation_around_x(0_degrees)'
      data.second = 'identity<4>()'
    end

    test_case do |data|
      data.first = 'rotation_around_y(0_degrees)'
      data.second = 'identity<4>()'
    end

    test_case do |data|
      data.first = 'rotation_around_z(0_degrees)'
      data.second = 'identity<4>()'
    end

    test_case do |data|
      data.first = 'rotation_around_x(0_degrees)'
      data.second = 'rotation_around_x(360_degrees)'
    end

    test_case do |data|
      data.first = 'rotation_around_x(-90_degrees)'
      data.second = 'rotation_around_x(270_degrees)'
    end

    test_case do |data|
      data.first = 'rotation_around_x(-180_degrees)'
      data.second = 'rotation_around_x(180_degrees)'
    end

    test_case do |data|
      data.first = 'scaling(-1,1,-1)'
      data.second = 'rotation_around_y(180_degrees)'
    end

    test_case do |data|
      data.first = 'scaling(-1,-1,1)'
      data.second = 'rotation_around_z(180_degrees)'
    end

    test_case do |data|
      data.first = 'scaling(1,-1,-1)'
      data.second = 'rotation_around_x(180_degrees)'
    end
  end


  test_suite do
    template do
      <<-END
        TEST_CASE("[Matrix] Matrix multiplication of #{first} and #{second}", "[Matrix]")
        {
            auto a = #{first};
            auto b = #{second};
            auto actual = a * b;
            auto expected = #{expected};

            REQUIRE( actual == approx(expected) );
        }
      END
    end

    test_case do |data|
      data.first = 'identity<4>()'
      data.second = 'identity<4>()'
      data.expected = 'identity<4>()'
    end

    test_case do |data|
      data.first = 'identity<4>()'
      data.second = 'scaling(2, 1, 1)'
      data.expected = 'scaling(2, 1, 1)'
    end

    test_case do |data|
      data.first = 'identity<4>()'
      data.second = 'scaling(1, 3, 1)'
      data.expected = 'scaling(1, 3, 1)'
    end

    test_case do |data|
      data.first = 'identity<4>()'
      data.second = 'scaling(1, 1, 4)'
      data.expected = 'scaling(1, 1, 4)'
    end

    test_case do |data|
      data.first = 'identity<4>()'
      data.second = 'rotation_around_x(40_degrees)'
      data.expected = 'rotation_around_x(40_degrees)'
    end

    test_case do |data|
      data.first = 'identity<4>()'
      data.second = 'rotation_around_y(10_degrees)'
      data.expected = 'rotation_around_y(10_degrees)'
    end

    test_case do |data|
      data.first = 'identity<4>()'
      data.second = 'rotation_around_z(250_degrees)'
      data.expected = 'rotation_around_z(250_degrees)'
    end

    test_case do |data|
      data.first = 'scaling(2, 1, 1)'
      data.second = 'scaling(2, 1, 1)'
      data.expected = 'scaling(4, 1, 1)'
    end

    test_case do |data|
      data.first = 'scaling(2, 1, 1)'
      data.second = 'scaling(0.5, 1, 1)'
      data.expected = 'scaling(1, 1, 1)'
    end

    test_case do |data|
      data.first = 'scaling(2, 1, 1)'
      data.second = 'scaling(1, 2, 1)'
      data.expected = 'scaling(2, 2, 1)'
    end

    test_case do |data|
      data.first = 'scaling(2, 1, 1)'
      data.second = 'scaling(1, 1, 2)'
      data.expected = 'scaling(2, 1, 2)'
    end

    test_case do |data|
      data.first = 'scaling(1, 3, 1)'
      data.second = 'scaling(1, 1, 2)'
      data.expected = 'scaling(1, 3, 2)'
    end

    test_case do |data|
      data.first = 'rotation_around_x(10_degrees)'
      data.second = 'rotation_around_x(10_degrees)'
      data.expected = 'rotation_around_x(20_degrees)'
    end

    test_case do |data|
      data.first = 'rotation_around_x(10_degrees)'
      data.second = 'rotation_around_x(-10_degrees)'
      data.expected = 'identity<4>()'
    end

    test_case do |data|
      data.first = 'rotation_around_x(10_degrees)'
      data.second = 'rotation_around_x(20_degrees)'
      data.expected = 'rotation_around_x(30_degrees)'
    end

    test_case do |data|
      data.first = 'rotation_around_y(10_degrees)'
      data.second = 'rotation_around_y(20_degrees)'
      data.expected = 'rotation_around_y(30_degrees)'
    end

    test_case do |data|
      data.first = 'rotation_around_z(10_degrees)'
      data.second = 'rotation_around_z(20_degrees)'
      data.expected = 'rotation_around_z(30_degrees)'
    end

    test_case do |data|
      data.first = 'translation(Vector3D(1,0,0))'
      data.second = 'translation(Vector3D(-1,0,0))'
      data.expected = 'identity<4>()'
    end

    test_case do |data|
      data.first = 'translation(Vector3D(0,5,0))'
      data.second = 'translation(Vector3D(0,-5,0))'
      data.expected = 'identity<4>()'
    end

    test_case do |data|
      data.first = 'translation(Vector3D(0,0,9))'
      data.second = 'translation(Vector3D(0,0,-9))'
      data.expected = 'identity<4>()'
    end

    test_case do |data|
      data.first = 'translation(Vector3D(1,0,0))'
      data.second = 'translation(Vector3D(0,0,2))'
      data.expected = 'translation(Vector3D(1,0,2))'
    end

    test_case do |data|
      data.first = 'translation(Vector3D(1,0,0))'
      data.second = 'Vector3D(0,0,0)'
      data.expected = 'Vector3D(0,0,0)'
    end

    test_case do |data|
      data.first = 'translation(Vector3D(0,1,0))'
      data.second = 'Vector3D(0,0,0)'
      data.expected = 'Vector3D(0,0,0)'
    end

    test_case do |data|
      data.first = 'translation(Vector3D(0,0,1))'
      data.second = 'Vector3D(0,0,0)'
      data.expected = 'Vector3D(0,0,0)'
    end

    test_case do |data|
      data.first = 'translation(Vector3D(1,0,0))'
      data.second = 'Point3D(0,0,0)'
      data.expected = 'Point3D(1,0,0)'
    end

    test_case do |data|
      data.first = 'translation(Vector3D(0,1,0))'
      data.second = 'Point3D(0,0,0)'
      data.expected = 'Point3D(0,1,0)'
    end

    test_case do |data|
      data.first = 'translation(Vector3D(0,0,1))'
      data.second = 'Point3D(0,0,0)'
      data.expected = 'Point3D(0,0,1)'
    end

    test_case do |data|
      data.first = 'scaling(2,1,1)'
      data.second = 'Point3D(0,0,0)'
      data.expected = 'Point3D(0,0,0)'
    end

    test_case do |data|
      data.first = 'scaling(2,1,1)'
      data.second = 'Point3D(1,0,0)'
      data.expected = 'Point3D(2,0,0)'
    end

    test_case do |data|
      data.first = 'scaling(2,3,4)'
      data.second = 'Point3D(1,1,1)'
      data.expected = 'Point3D(2,3,4)'
    end

    test_case do |data|
      data.first = 'scaling(2,3,4)'
      data.second = 'Point3D(2,2,2)'
      data.expected = 'Point3D(4,6,8)'
    end

    test_case do |data|
      data.first = 'rotation_around_x(180_degrees)'
      data.second = 'Point3D(2,2,2)'
      data.expected = 'Point3D(2,-2,-2)'
    end

    test_case do |data|
      data.first = 'rotation_around_y(180_degrees)'
      data.second = 'Point3D(2,2,2)'
      data.expected = 'Point3D(-2,2,-2)'
    end

    test_case do |data|
      data.first = 'rotation_around_z(180_degrees)'
      data.second = 'Point3D(2,2,2)'
      data.expected = 'Point3D(-2,-2,2)'
    end

    test_case do |data|
      data.first = 'rotation_around_x(180_degrees)'
      data.second = 'Vector3D(2,2,2)'
      data.expected = 'Vector3D(2,-2,-2)'
    end

    test_case do |data|
      data.first = 'rotation_around_y(180_degrees)'
      data.second = 'Vector3D(2,2,2)'
      data.expected = 'Vector3D(-2,2,-2)'
    end

    test_case do |data|
      data.first = 'rotation_around_z(180_degrees)'
      data.second = 'Vector3D(2,2,2)'
      data.expected = 'Vector3D(-2,-2,2)'
    end
    
    test_case do |data|
      data.first = 'rotation_around_x(90_degrees)'
      data.second = 'Point3D(0,0,0)'
      data.expected = 'Point3D(0,0,0)'
    end

    test_case do |data|
      data.first = 'rotation_around_x(90_degrees)'
      data.second = 'Point3D(1,0,0)'
      data.expected = 'Point3D(1,0,0)'
    end

    test_case do |data|
      data.first = 'rotation_around_x(90_degrees)'
      data.second = 'Point3D(0,1,0)'
      data.expected = 'Point3D(0,0,1)'
    end

    test_case do |data|
      data.first = 'rotation_around_x(90_degrees)'
      data.second = 'Point3D(0,0,1)'
      data.expected = 'Point3D(0,-1,0)'
    end

    test_case do |data|
      data.first = 'rotation_around_y(90_degrees)'
      data.second = 'Point3D(0,0,0)'
      data.expected = 'Point3D(0,0,0)'
    end

    test_case do |data|
      data.first = 'rotation_around_y(90_degrees)'
      data.second = 'Point3D(0,1,0)'
      data.expected = 'Point3D(0,1,0)'
    end

    test_case do |data|
      data.first = 'rotation_around_y(90_degrees)'
      data.second = 'Point3D(1,0,0)'
      data.expected = 'Point3D(0,0,-1)'
    end

    test_case do |data|
      data.first = 'rotation_around_y(90_degrees)'
      data.second = 'Point3D(0,0,-1)'
      data.expected = 'Point3D(-1,0,0)'
    end

    test_case do |data|
      data.first = 'rotation_around_y(90_degrees)'
      data.second = 'Point3D(-1,0,0)'
      data.expected = 'Point3D(0,0,1)'
    end

    test_case do |data|
      data.first = 'rotation_around_z(90_degrees)'
      data.second = 'Point3D(0,0,0)'
      data.expected = 'Point3D(0,0,0)'
    end

    test_case do |data|
      data.first = 'rotation_around_z(90_degrees)'
      data.second = 'Point3D(0,0,1)'
      data.expected = 'Point3D(0,0,1)'
    end

    test_case do |data|
      data.first = 'rotation_around_z(90_degrees)'
      data.second = 'Point3D(1,0,0)'
      data.expected = 'Point3D(0,1,0)'
    end

    test_case do |data|
      data.first = 'rotation_around_z(90_degrees)'
      data.second = 'Point3D(0,1,0)'
      data.expected = 'Point3D(-1,0,0)'
    end
  end
end
